### PR TITLE
systemctl: do not segfault when we cannot find template unit

### LINKED ISSUE #4912: systemctl segfault when told to cat a non-existent template unit
```
Core was generated by `systemctl cat test@.target test@.service'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  strrchr () at ../sysdeps/x86_64/strrchr.S:32
32              movdqu  (%rdi), %xmm0
(gdb) bt
#0  strrchr () at ../sysdeps/x86_64/strrchr.S:32
#1  0x00007f3da9b357fe in __GI___basename (filename=0x0) at basename.c:24
#2  0x0000557b3df2fd91 in unit_find_paths (bus=0x557b3e9d6a90, unit_name=0x557b3e9d6740 "test@.target", lp=0x7ffccb170830, fragment_path=0x7ffccb170810, dropin_paths=0x7ffccb170818) at src/systemctl/systemctl.c:2584
#3  0x0000557b3df3aae5 in cat (argc=3, argv=0x7ffccb170aa0, userdata=0x0) at src/systemctl/systemctl.c:5324
#4  0x00007f3daa111c6b in dispatch_verb (argc=4, argv=0x7ffccb170a98, verbs=0x557b3df50c60 <verbs>, userdata=0x0) at src/basic/verbs.c:92
#5  0x0000557b3df4377f in systemctl_main (argc=4, argv=0x7ffccb170a98) at src/systemctl/systemctl.c:8141
#6  0x0000557b3df44572 in main (argc=4, argv=0x7ffccb170a98) at src/systemctl/systemctl.c:8412
```

Regression introduced in c6dd36b65caa9acc03a8e77b7db752574cf36eb4.

@fsateler

### PR DESCRIPTION
**Pull request by keszybz**
```
Core was generated by `systemctl cat test@.target test@.service'.
Program terminated with signal SIGSEGV, Segmentation fault.
32              movdqu  (%rdi), %xmm0
(gdb) bt
-0  strrchr () at ../sysdeps/x86_64/strrchr.S:32
-1  0x00007f57fdf837fe in __GI___basename (filename=0x0) at basename.c:24
-2  0x000055b8a77d0d91 in unit_find_paths (bus=0x55b8a9242f90, unit_name=0x55b8a92428f0 "test@.service", lp=0x7ffdc9070400, fragment_path=0x7ffdc90703e0, dropin_paths=0x7ffdc90703e8) at src/systemctl/systemctl.c:2584
-3  0x000055b8a77dbae5 in cat (argc=3, argv=0x7ffdc9070678, userdata=0x0) at src/systemctl/systemctl.c:5324
-4  0x00007f57fe55fc6b in dispatch_verb (argc=5, argv=0x7ffdc9070668, verbs=0x55b8a77f1c60 <verbs>, userdata=0x0) at src/basic/verbs.c:92
-5  0x000055b8a77e477f in systemctl_main (argc=5, argv=0x7ffdc9070668) at src/systemctl/systemctl.c:8141
-6  0x000055b8a77e5572 in main (argc=5, argv=0x7ffdc9070668) at src/systemctl/systemctl.c:8412
```

The right behaviour is not easy in this case. Implement some "sensible" logic.

Fixes #4912.

I'm not sure if this is the right solution…

### CODE REVIEW

File: src/systemctl/systemctl.c [hunk 1]
```diff
@@ -2581,7 +2580,22 @@ static int unit_find_paths(
                         }
                 }
 
-                r = set_put(names, basename(path));
+                if (path)
+                        /* We found the unit file. If we followed symlinks, this name might be
+                         * different then the unit_name with started with. Look for dropins matching
```
Comment (arvidjaar) [inline src/systemctl/systemctl.c:2585]: s/then/than/ here?

File: src/systemctl/systemctl.c [hunk 2]
```diff
@@ -2581,7 +2580,22 @@ static int unit_find_paths(
...
+                         * that "final" name. */
+                        r = set_put(names, basename(path));
+                else if (!template)
+                        /* No unit file, let's look for dropins matching the original name.
+                         * systemd has fairly complicated rules (based on unit type and provenience),
```
Comment (arvidjaar) [inline src/systemctl/systemctl.c:2590]: I had to look "provenience" in dictionary and even then I am not sure I got it right. May be something more simple?

Comment (poettering) [reply src/systemctl/systemctl.c:2590]: but it's the usual technical term used in cases like this, hence I think it's good the way it is...

**Comment by poettering** [comment]
patch looks good to me... the "than"/"then" issue, doesn't appear reason enough not to merge this ;-)

**Comment by fsateler** [comment]
Sorry I did not find time to take a look. Thanks for the fix up @keszybz !

**Pull request merged**